File: 2020/day05/sol.rs
use std::collections::HashSet;
use std::fs;

fn main() {
    let input = fs::read_to_string("input").expect("Can't read puzzle input file.");

    let seat_ids: Vec<_> = input.lines().map(seat_id).collect();
    let max_seat_id = seat_ids.iter().by_ref().max().unwrap();

    println!("part 1: {}", max_seat_id);

    let seats: HashSet<_> = seat_ids.iter().cloned().collect();
    let min_seat_id = seat_ids.iter().by_ref().min().unwrap();

    (min_seat_id + 1..max_seat_id - 1).for_each(|i| {
        if seats.contains(&(i - 1)) && !seats.contains(&i) && seats.contains(&(i + 1)) {
            println!("part 2: {}", i);
        }
    });
}

fn seat_id(boarding_pass: &str) -> u32 {
    let mut boarding_pass_chars = boarding_pass.chars();
    let row: u32 = boarding_pass_chars
        .by_ref()
        .take(7)
        .enumerate()
        .map(|(i, c)| match c {
            'B' => 1 << (6 - i),
            _ => 0,
        })
        .sum();

    let column: u32 = boarding_pass_chars
        .take(3)
        .enumerate()
        .map(|(i, c)| match c {
            'R' => 1 << (2 - i),
            _ => 0,
        })
        .sum();

    row * 8 + column
}

File: 2020/day05/sol.cpp
#include <fstream>
#include <iostream>
#include <map>

int main() {
    std::ifstream input("input");

    int max = 0;
    std::map<int, bool> seats;

    std::string line;
    while (std::getline(input, line)) {
        int row_lower = 0;
        int column_lower = 0;

        for (int j = 0; j < 7; ++j) {
            int m = 6 - j;
            if (line[j] == 'B') {
                row_lower += 1 << m;
            }
        }

        for (int j = 7; j < 10; ++j) {
            int m = 9 - j;
            if (line[j] == 'R') {
                column_lower += 1 << m;
            }
        }

        int seat_id = row_lower * 8 + column_lower;
        if (seat_id > max) {
            max = seat_id;
        }

        seats[seat_id] = true;
    }

    std::cout << "part 1: " << max << '\n';

    for (int id = 1; id < 128 * 8 - 2; ++id) {
        if (seats[id - 1] && !seats[id] && seats[id + 1]) {
            std::cout << "part 2: " << id << '\n';
            return 0;
        }
    }

    return 0;
}
